giraffe/src/utils/mosaicTooltip.ts

```diff
@@ -0,0 +1,115 @@
+import {range} from 'd3-array'
+
+import {getRangeLabel} from './tooltip'
+import {X_MIN, X_MAX, FILL, SERIES} from '../constants/columnKeys'
+import {
+  Table,
+  Scale,
+  TooltipData,
+  TooltipColumn,
+  ColumnGroupMap,
+} from '../types'
+
+export const findHoveredBoxes = (
+  boxTable: Table,
+  hoverX: number | null,
+  hoverY: number | null,
+  xScale: Scale<number, number>,
+  yScale: Scale<number, number>,
+  yDomain: number[]
+): number[] => {
+  if (!hoverX || !hoverY) {
+    return []
+  }
+  const xMinData = boxTable.getColumn(X_MIN, 'number')
+  const xMaxData = boxTable.getColumn(X_MAX, 'number')
+  const dataX = xScale.invert(hoverX)
+  const dataY = yScale.invert(hoverY)
+  // Find all boxes whose x extent contain the mouse x position
+  const xIndices = range(0, xMinData.length).filter(
+    i => xMinData[i] <= dataX && xMaxData[i] > dataX
+  )
+
```

Comment:
first things first. `xIndicies` is referenced directly in the return statement of this function:

```ts
  if (xIndices) {
    xyIndices = [xIndices[finalXIndex]]
  } else {
    xyIndices = []
  }

   return xyIndices
```

if this comes back as falsy, we can just early return from this function right now without any more looping.

```suggestion
if (!xIndices) {
  return []
}
```

now, any code below this `if` statement (which i've heard referred to as guard statements) is considered happy path - it can be assumed to work, since we checked an exit condition early. this pattern (where we check if there are any conditions that will cause our algorithm to fail, and if so, bail) is called an early return.

this early return can drastically change the looping code below, because now loops can return early as well, rather than maintaining state in a variable, and handling a return based on that (see below)